### PR TITLE
Linlin

### PR DESCRIPTION
**Pull request by Carolyn1204**

### CODE REVIEW

File: server/src/routes/order.route.ts
Comment (XintangZ) [inline server/src/routes/order.route.ts:1]: 1. all order routs should use jwt auth middleware (the user should login to perform these actions)
2. suggested endpoint design:

	// Create a new order
	router.post('/orders', authenticateJWT, createOrder);

	// Get all orders
	router.get('/orders', authenticateJWT, getAllOrders);

	// Get a specific order by ID
	router.get('/orders/:id', authenticateJWT, getOrder);

	// Update an order by ID
	router.put('/orders/:id', authenticateJWT, updateOrder);

File: server/src/schema/order.schema.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { z } from 'zod';
+import { Types } from 'mongoose';
+
+// Zod validation schema for Order
+export const createOrderSchema = z.object({
+
+    userId: z.string().min(1, { message: 'Invalid user ID format' }),
```
Comment (XintangZ) [inline server/src/schema/order.schema.ts:7]: userId can be obtained from decoded jwt

File: server/src/controllers/order.controller.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
+import { Request, Response } from 'express';
+import { Order } from '../models/index';
+import { createOrderSchema } from '../schema';
+import mongoose from 'mongoose';
+import { z } from 'zod';
+
+// Create a new order
+export const createOrder = async (req: Request, res: Response) => {
+    try {
```
Comment (XintangZ) [inline server/src/controllers/order.controller.ts:9]: get userId from decoded jwt

File: server/src/controllers/order.controller.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        if (error instanceof z.ZodError) {
+            return res.status(400).json({ error: error.errors });
+        }
+        return res.status(400).json({ error: error instanceof Error ? error.message : 'An unexpected error occurred' });
+    }
+};
+
+// Get all orders
+export const getAllOrders = async (req: Request, res: Response) => {
+    const { page = 1, limit = 10 } = req.query;
```
Comment (XintangZ) [inline server/src/controllers/order.controller.ts:26]: 1. check if the decoded user is admin. yes->2. no->3.
2. check if the request comes with a query param (/orders?userId=blablabla), yes->4. no->5.
3. return all orders matching the decoded userId (the current logged in user)
4. return all orders matching the provided userId
5. return all orders in the db

File: server/src/controllers/order.controller.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+export const getOrder = async (req: Request, res: Response) => {
+    const { id } = req.params;
+    if (!mongoose.isValidObjectId(id)) {
+        return res.status(400).json({ error: 'Invalid order ID format' });
+    }
+    try {
+        const order = await Order.findById(id);
+        if (!order) {
+            return res.status(404).json({ error: 'Order not found' });
+        }
```
Comment (XintangZ) [inline server/src/controllers/order.controller.ts:118]: 1. after the order is found
2. check if the decoded user is admin. yes->2. no->3.
3. return the found order 
4. check if the found order's userId matches the decoded userId, yes->3. no->5.
5. return 403 forbidden

File: server/src/controllers/order.controller.ts (hunk 2 above)
Comment (XintangZ) [inline server/src/controllers/order.controller.ts:9]: get userId from decoded jwt

File: server/src/controllers/order.controller.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            return res.status(400).json({ error: error.errors });
+        }
+        return res.status(400).json({ error: error instanceof Error ? error.message : 'An unexpected error occurred' });
+    }
+};
+
+// Get all orders
+export const getAllOrders = async (req: Request, res: Response) => {
+    const { page = 1, limit = 10 } = req.query;
+
```
Comment (XintangZ) [inline server/src/controllers/order.controller.ts:45]: 1. check if the decoded user is admin. yes->2. no->3.
2. check if the request comes with a query param (/orders?userId=blablabla), yes->4. no->5.
3. return all orders matching the decoded userId (the current logged in user)
4. return all orders matching the provided userId
5. return all orders in the db

File: server/src/controllers/order.controller.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        return res.status(400).json({ error: 'Invalid order ID format' });
+    }
+
+    // Validate status value
+    if (!['Pending', 'Shipped', 'Cancelled'].includes(status)) {
+        return res.status(400).json({ error: 'Invalid status value' });
+    }
+
+    try {
+        const updatedOrder = await Order.findByIdAndUpdate(
```
Comment (XintangZ) [inline server/src/controllers/order.controller.ts]: 1. find the order
2. check if the status from the request is 'cancel'. yes->3. no->4.
3. update the order, return
4. check if the decoded user is admin. yes->3. no->5.
5. return 403 forbidden

File: server/src/routes/order.route.ts
Comment (XintangZ) [inline server/src/routes/order.route.ts:1]: 1. all order routs should use jwt auth middleware (the user should login to perform these actions)
2. suggested endpoint design:

	// Create a new order
	router.post('/orders', authenticateJWT, createOrder);

	// Get all orders
	router.get('/orders', authenticateJWT, getAllOrders);

	// Get a specific order by ID
	router.get('/orders/:id', authenticateJWT, getOrder);

	// Update an order by ID
	router.put('/orders/:id', authenticateJWT, updateOrder);

File: server/src/schema/order.schema.ts (hunk 1 above)
Comment (XintangZ) [inline server/src/schema/order.schema.ts]: userId can be obtained from decoded jwt

**Review by XintangZ — CHANGES_REQUESTED** [review]

**Review by XintangZ — APPROVED** [review]

**Review by XintangZ — APPROVED** [review]

**Review by aqizzz — APPROVED** [review]
LGTM

**Pull request merged**